### PR TITLE
[image-picker][ios] Support selecting multiple media

### PR DESCRIPTION
**Pull request by barthap**
# Why

We want to support multi-select on all platforms

Part of ENG-2795,
Part of ENG-5617
Part of module audit, frequently requested on twitter.

# How

On iOS 14+ there is a [PHPickerViewController](https://developer.apple.com/documentation/photokit/phpickerviewcontroller) which supports multiple selection. **It does not support cropping, though**. We now use the new picker only if `iOS >= 14 && allowsMultipleSelection == true`.

Both configuration and image/video handling differ significantly for this new picker, so I created a separate set of methods to support the new flow.

Currently supported features are:
- Select images/videos/all
- compression quality settings (for JPEGs and others)
- EXIF
- base64

#### This PR implements basic support, in the future it can be expanded further:
- Support selection limit (e.g. up to 5 images)
- Return asset IDs (they could then be queried using expo-media-library)
- Support initial selection (provide asset IDs at start, they become preselected)
- Support ordered selection (assets are returned in the exact order user was selecting them - there are numbers on selection bages then)
- Restore support of animated gifs (they're now converted to jpg)

# Test Plan

NCL

https://user-images.githubusercontent.com/278340/177273376-e5d892a5-bc70-409e-9a05-113e320fb424.mov

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by lukmccall — APPROVED** [review]

**Review by tsapeta — APPROVED** [review]

File: packages/expo-image-picker/ios/ImagePickerModule.swift [hunk 1]
```diff
@@ -139,6 +163,31 @@ public class ImagePickerModule: Module, OnMediaPickingResultHandler {
     self.currentPickingContext = nil
   }
 
+  @available(iOS 14, *)
+  func didPickMultipleMedia(selection: [PHPickerResult]) {
+    guard let options = self.currentPickingContext?.options,
+          let promise = self.currentPickingContext?.promise else {
+      NSLog("Picking operation context has been lost.")
```
[the remark this answers is not in the document]
Comment (barthap) [reply packages/expo-image-picker/ios/ImagePickerModule.swift:170]: Shouldn't each expo-module have its own Logger category?

File: packages/expo-image-picker/ios/ImagePickerModule.swift [hunk 2]
```diff
@@ -89,6 +90,16 @@ public class ImagePickerModule: Module, OnMediaPickingResultHandler {
...
+
+    let pickingContext = PickingContext(promise: promise,
+                                        options: options,
+                                        imagePickerHandler: imagePickerDelegate)
+
+    if #available(iOS 14, *), options.allowsMultipleSelection && !options.allowsEditing && sourceType != .camera {
+      self.launchMultiSelectPicker(pickingContext: pickingContext, currentViewController: currentViewController)
+      return
+    }
+
```
Comment (bbarthec) [inline packages/expo-image-picker/ios/ImagePickerModule.swift:102]: Ok, I see that this function is growing in logic and I kind of dislike it 😅 
I suggest making this function a very top level and push the logic to other extracted functions 🤓 
I would like to ask you to provide something like this:
```swift
return switch (options) {
.case shouldLaunchMultiselectionPicker -> lauchMultiselectorPicker(...some options)
.case shouldLaunchLegacyPicker -> launchLegacyPicker(...some options)
.case shouldLaunchCamera -> launchCamera(...some options)
}
```

I guess some of `launch*` methods might share common logic:
```swift
private func launchLegacyPicker() {
  let picker = UIImagePickerController(options) // use can use https://docs.swift.org/swift-book/LanguageGuide/Extensions.html
  
  // add some specific logic
  picker.sourceType = .photoLibrary
  // etc

  launchPicker(picker)
}

private func launchCamera() {
  let picker = UIImagePickerController(options) // use can use https://docs.swift.org/swift-book/LanguageGuide/Extensions.html
  
  // add some specific logic
  picker.sourceType = .camera
  // etc

  launchPicker(picker)
}
```
When I look at this suggestion of mine I come to the conclusion that `.camera` and `.legacyPicker` are almost the same 😅

File: packages/expo-image-picker/ios/MediaHandler.swift [hunk 3]
```diff
@@ -97,6 +182,37 @@ internal struct MediaHandler {
...
+        let result = VideoInfo(uri: targetUrl.absoluteString,
+                               width: size.width,
+                               height: size.height,
+                               duration: duration)
+        completion(.success(result))
+      } catch let exception as Exception {
+        return completion(.failure(exception))
+      } catch {
+        return completion(.failure(UnexpectedException(error)))
+      }
```
Comment (bbarthec) [inline packages/expo-image-picker/ios/MediaHandler.swift:212]: This looks similar to the other `handleVideo` function body. Unify it.

File: packages/expo-image-picker/ios/MediaHandler.swift [hunk 4]
```diff
@@ -62,6 +108,45 @@ internal struct MediaHandler {
...
+                               width: image.size.width,
+                               height: image.size.height,
+                               base64: base64,
+                               exif: exif)
+        completion(.success(result))
+      } catch let exception as Exception {
+        return completion(.failure(exception))
+      } catch {
+        return completion(.failure(UnexpectedException(error)))
+      }
```
Comment (bbarthec) [inline packages/expo-image-picker/ios/MediaHandler.swift:146]: This looks the same os the other `handleImage` function body. Unify it.

**Review by bbarthec — CHANGES_REQUESTED** [review]
Overall good job on adding this new picker 👍 
I appreciate you've followed my code concepts I've wrote when I was working with this module 😉 
My remarks/questions/arguments:
- let's try not to create any duplication and unify all logic in some atomic functions, e.g:
```swift
picker.delegate = pickingContext.imagePickerHandler
    picker.presentationController?.delegate = pickingContext.imagePickerHandler

    // Store picking context as we're navigating to the different view controller (starting asynchronous flow)
    self.currentPickingContext = pickingContext
    currentViewController.present(picker, animated: true, completion: nil)
```
- let's refactor too complicated functions (like `launchImagePicker`) into multiple atomic ones
- does iOS 14 picker look any different then the legacy one? If so I'd like to have an option to select single image using this new picker as well (additional param in options?) or default to this new picker and have an option to use the legacy one 🤔

File: packages/expo-image-picker/ios/MediaHandler.swift (hunk 3 above)
Comment (barthap) [reply packages/expo-image-picker/ios/MediaHandler.swift:204]: There are subtle differences in instructions inside `handleVideo` and `handleImage` - trials of unifying this introduced even more mess in helper functions so imo it's not worth it for now.

File: packages/expo-image-picker/ios/ImagePickerModule.swift (hunk 2 above)
Comment (barthap) [reply packages/expo-image-picker/ios/ImagePickerModule.swift]: Cleaned it up a bit, but left camera and legacy picker together for now.

**Comment by barthap** [comment]
> does iOS 14 picker look any different then the legacy one? If so I'd like to have an option to select single image using this new picker as well (additional param in options?) or default to this new picker and have an option to use the legacy one

They look and identical when `selectionLimit == 1` - the difference is in permissions and configuration options - new picker doesn't require permissions, but has less configuration options (video export preset, video quality etc) - this will need to be implemented manually.

File: packages/expo-image-picker/src/ImagePicker.ts [hunk 5]
```diff
@@ -229,6 +229,13 @@ export async function launchImageLibraryAsync<T extends ImagePickerOptions>(
   if (!ExponentImagePicker.launchImageLibraryAsync) {
     throw new UnavailabilityError('ImagePicker', 'launchImageLibraryAsync');
   }
+  if (options?.allowsEditing && options.allowsMultipleSelection) {
+    console.warn(
+      '[expo-image-picker] `allowsEditing` is not supported when `allowsMultipleSelection` is enabled and would be ignored.' +
```
Comment (bbarthec) [inline packages/expo-image-picker/src/ImagePicker.ts:234]:
```suggestion
      '[expo-image-picker] `allowsEditing` is not supported when `allowsMultipleSelection` is enabled and will be ignored.' +
```
```suggestion
      '[expo-image-picker] `allowsEditing` is not supported when `allowsMultipleSelection` is enabled and would be ignored.' +
```

**Review by bbarthec — APPROVED** [review]
Ok, I'm fine with the current shape of this PR 👍 
Thank you for taking care of my concerns 😉 
I looking forward to the next PRs enriching this module even further 😉

**Comment by GleidsonDaniel** [comment]
This PR generates a bug that when selecting a video and an image using this method it takes a long time until it responds with the data, more than 10 seconds.

**Pull request merged**